tools/httpbintest/src/lib.rs

```diff
@@ -0,0 +1,1304 @@
+    pub fn set_concurrency(&mut self, concurrent: bool) {
+        self.concurrent = concurrent;
+    }
+
+    /// Returns the total number of requests in a test.
+    pub fn requests_count(&mut self) -> usize {
+        self.reqs.len()
+    }
+
+    pub fn endpoint(&self) -> url::Url {
+        self.endpoint.clone()
+    }
+
+    /// Send one or more requests based on test type and the concurrency
+    /// property. If any send fails, a quiche::h3::Error is returned.
+    pub fn send_requests(
+        &mut self, conn: &mut quiche::Connection,
+        h3_conn: &mut quiche::h3::Connection,
+    ) -> quiche::h3::Result<()> {
+        if self.reqs.len() - self.current_idx == 0 {
+            return Err(quiche::h3::Error::Done);
+        }
+
+        let reqs_to_make = if self.concurrent {
+            self.reqs.len() - self.current_idx
+        } else {
+            1
+        };
+
+        for _ in 0..reqs_to_make {
+            let req = &self.reqs[self.current_idx];
+
+            info!("sending HTTP request {:?}", req.hdrs);
+
+            let s =
+                match h3_conn.send_request(conn, &req.hdrs, req.body.is_none()) {
+                    Ok(stream_id) => stream_id,
+
+                    Err(e) => {
+                        error!("failed to send request {:?}", e);
+                        return Err(e);
+                    },
+                };
+
+            self.issued_reqs.insert(s, self.current_idx);
+
+            if let Some(body) = &req.body {
+                info!("sending body {:?}", body);
+
+                if let Err(e) = h3_conn.send_body(conn, s, body, true) {
+                    error!("failed to send request body {:?}", e);
+                    return Err(e);
+                }
+            }
+
+            self.current_idx += 1;
+        }
+
+        Ok(())
+    }
+
+    /// Append response headers for an issued request.
+    pub fn add_response_headers(&mut self, stream_id: u64, headers: &[Header]) {
+        let i = self.issued_reqs.get(&stream_id).unwrap();
+        self.reqs[*i].resp_hdrs.extend_from_slice(headers);
+    }
+
+    /// Append data to the response body for an issued request.
+    pub fn add_response_body(
+        &mut self, stream_id: u64, data: &[u8], data_len: usize,
+    ) {
+        let i = self.issued_reqs.get(&stream_id).unwrap();
+        self.reqs[*i].resp_body.extend_from_slice(&data[..data_len]);
+    }
+
+    /// Execute the test assertion(s).
+    pub fn assert(&mut self) -> bool {
+        (self.assert)(&self.reqs)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use std::sync::Once;
+
+    static INIT: Once = Once::new();
+
+    fn run_test<T>(test: T) -> ()
+    where
+        T: FnOnce() -> () + std::panic::UnwindSafe,
+    {
+        INIT.call_once(|| {
+            env_logger::builder()
+                .default_format_timestamp_nanos(true)
+                .init()
+        });
+
+        let result = std::panic::catch_unwind(|| test());
```

Comment:
Instead of catching panics, it would probably be better to just let them crash the program, so we would get better debug output when a test fails (see also the other comments about using real asserts).

Reply:
I was copying the link you suggested - https://medium.com/@ericdreichert/test-setup-and-teardown-in-rust-without-a-framework-ba32d97aa5ab

I guess it depends if there is any expensive setup that needs to be tidied up in the case of failure but I don't see a strong need for that in the case of this client-oriented testing code.